File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.py
def solution(land):
    n = len(land)        # 세로
    m = len(land[0])     # 가로

    visited = [[False] * m for _ in range(n)]
    col_total = [0] * m  # 각 열마다 얻을 수 있는 석유량
    directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

    for i in range(n):
        for j in range(m):
            # 아직 방문하지 않은 석유 칸이면 새로운 덩어리 탐색 시작
            if land[i][j] == 1 and not visited[i][j]:
                stack = [(i, j)]
                visited[i][j] = True

                size = 0          # 이 덩어리의 전체 크기
                cols = set()      # 이 덩어리가 포함하는 열 인덱스들

                # DFS로 연결된 모든 석유 칸 탐색
                while stack:
                    x, y = stack.pop()
                    size += 1
                    cols.add(y)   # 열 기록

                    for dx, dy in directions:
                        nx, ny = x + dx, y + dy
                        if 0 <= nx < n and 0 <= ny < m:
                            if not visited[nx][ny] and land[nx][ny] == 1:
                                visited[nx][ny] = True
                                stack.append((nx, ny))

                # 이 덩어리가 지나가는 각 열마다 덩어리 크기를 더한다
                for c in cols:
                    col_total[c] += size

    return max(col_total)

File: 프로그래머스/2/250136. ［PCCP 기출문제］ 2번 ／ 석유 시추/［PCCP 기출문제］ 2번 ／ 석유 시추.js
function solution(land) {
    const n = land.length;        // 세로
    const m = land[0].length;     // 가로

    const visited = Array.from({ length: n }, () => Array(m).fill(false));
    const colTotal = Array(m).fill(0);   // 각 열에서 얻을 수 있는 석유량

    const dr = [1, -1, 0, 0];
    const dc = [0, 0, 1, -1];

    for (let i = 0; i < n; i++) {
        for (let j = 0; j < m; j++) {
            // 새로운 석유 덩어리 발견
            if (land[i][j] === 1 && !visited[i][j]) {
                const stack = [[i, j]];
                visited[i][j] = true;

                let size = 0;            // 덩어리 크기
                const cols = new Set();  // 이 덩어리가 차지하는 열들

                // DFS (스택)
                while (stack.length > 0) {
                    const [r, c] = stack.pop();
                    size++;
                    cols.add(c);  // 이 열을 지나감

                    for (let k = 0; k < 4; k++) {
                        const nr = r + dr[k];
                        const nc = c + dc[k];

                        if (nr < 0 || nr >= n || nc < 0 || nc >= m) continue;
                        if (!visited[nr][nc] && land[nr][nc] === 1) {
                            visited[nr][nc] = true;
                            stack.push([nr, nc]);
                        }
                    }
                }

                // 이 덩어리가 지나가는 모든 열에 덩어리 크기 더하기
                for (const c of cols) {
                    colTotal[c] += size;
                }
            }
        }
    }

    return Math.max(...colTotal);
}